File: shopping_cart.js
// Import propt-sync for User-Input
const prompt = require("prompt-sync")();

// declaring the discount amount to null
let flat_10_discount_amount = null;
let bulk_5_discount_amount = null;
let bulk_10_discount_amount = null;
let tiered_50_discount_amount = null;

//  discount 1 - flat_10_discount
const flat_10_discount = () => {
    return 10.0;
};
// discount 2 - bulk_5_discount
const bulk_5_discount = (priceOfProduct) => {
    return parseFloat(priceOfProduct * (5 / 100));
};
// discount 3 - bulk_10_discount
const bulk_10_discount = (cartTotal) => {
    return parseFloat(cartTotal * (10 / 100));
};
// discount 4 - tiered_50_discount
const tiered_50_discount = (priceOfDiscountProduct) => {
    return parseFloat(priceOfDiscountProduct * (50/100));
};

const applyDiscount = () => {
    const discountAmounts = new Map();
    //  check flat_10_discount
    if (cartTotal > 200){
        flat_10_discount_amount = flat_10_discount();
        discountAmounts.set("flat_10_discount",flat_10_discount_amount);
    }
    // check bulk_5_discount
    if((quantityOfProductA > 10) || (quantityOfProductB > 10) || (quantityOfProductC > 10)){
        let bulk_5_discount_amount_A = null;
        let bulk_5_discount_amount_B = null;
        let bulk_5_discount_amount_C = null;
        if(quantityOfProductA > 10){
            bulk_5_discount_amount_A = bulk_5_discount(totalAmountOfProductA);
        }
        if(quantityOfProductB > 10){
            bulk_5_discount_amount_B = bulk_5_discount(totalAmountOfProductB);
        }
        if(quantityOfProductC > 10){
            bulk_5_discount_amount_C = bulk_5_discount(totalAmountOfProductC);
        }
        bulk_5_discount_amount = bulk_5_discount_amount_A + bulk_5_discount_amount_B + bulk_5_discount_amount_C;
        discountAmounts.set("bulk_5_discount", bulk_5_discount_amount);
    }
    // check bulk_10_discount
    if(totalQuantity > 20){
        bulk_10_discount_amount = bulk_10_discount(cartTotal);
        discountAmounts.set("bulk_10_discount", bulk_10_discount_amount);
    }
    // check tiered_50_discount
    if(totalQuantity > 30){
        let tiered_50_discount_amount_A = null;
        let tiered_50_discount_amount_B = null;
        let tiered_50_discount_amount_C = null;
        if(quantityOfProductA > 15){
            let remainingQuantity = quantityOfProductA - 15;
            tiered_50_discount_amount_A = (remainingQuantity * catalog.get("Product A")) - tiered_50_discount(remainingQuantity * catalog.get("Product A"));
        }
        if(quantityOfProductB > 15){
            let remainingQuantity = quantityOfProductB - 15;
            tiered_50_discount_amount_B = (remainingQuantity * catalog.get("Product B")) - tiered_50_discount(remainingQuantity * catalog.get("Product B"));
        }
        if(quantityOfProductC > 15){
            let remainingQuantity = quantityOfProductC - 15;
            tiered_50_discount_amount_C = (remainingQuantity * catalog.get("Product C")) - tiered_50_discount(remainingQuantity * catalog.get("Product C"));
        }
        tiered_50_discount_amount = tiered_50_discount_amount_A + tiered_50_discount_amount_B + tiered_50_discount_amount_C;
        discountAmounts.set("tiered_50_discount" ,tiered_50_discount_amount);
    }
    return discountAmounts;
};

// Minimum Discount amount is beneficial to the customer
const findBeneficialDiscount = (discountAmounts) => {
    let maxPrice = 0;
    let discountNameOfmaxPrice = null;
    for(const [name, price] of discountAmounts){
        if(price > maxPrice){
            maxPrice = price;
            discountNameOfmaxPrice = name;
        }
    }
    return {discountName : discountNameOfmaxPrice, discountPrice : maxPrice};
}

const shipping = () => {
    // 10units = 1 package
    let unitsPerPackage = 10
    let package = Math.floor(totalQuantity / unitsPerPackage);
    let remainingUnits = totalQuantity % unitsPerPackage;
    if((remainingUnits > 0) && (remainingUnits <= unitsPerPackage)){
        package = package + 1
    }
    let shippingFee = package * 5.0
    return shippingFee
};

const giftWrap = () => {
    if(isWrapped == "yes"){
        wrapAmount = totalQuantity * 1.0;
    }
    else if(isWrapped == "no"){
        wrapAmount = 0.0;
    }
    return wrapAmount
};

const display = (discountAmounts) => {
    console.log("\nProduct Name\tQuantity\tTotal Amount");
    console.log(`Product A\t${quantityOfProductA}\t\t$${totalAmountOfProductA.toFixed(2)}`);
    console.log(`Product B\t${quantityOfProductB}\t\t$${totalAmountOfProductB.toFixed(2)}`);
    console.log(`Product C\t${quantityOfProductC}\t\t$${totalAmountOfProductC.toFixed(2)}`);
    console.log(`\nSubTotal                : $${cartTotal.toFixed(2)}`);
    const beneficialDiscount = findBeneficialDiscount(discountAmounts);
    console.log(`Discount Name Appliced  : ${beneficialDiscount.discountName}`);
    console.log(`Discount Amount         : $${beneficialDiscount.discountPrice.toFixed(2)}`);
    const shipFee = shipping();
    const giftWrapFee = giftWrap();
    console.log(`Shipping Fee            : $${shipFee.toFixed(2)}`);
    console.log(`Gift Wrap Fee           : $${giftWrapFee.toFixed(2)}`);
    const totalAmount = cartTotal - beneficialDiscount.discountPrice + shipFee + giftWrapFee
    console.log(`Total                   : $${totalAmount.toFixed(2)}`)
};

// Defining a map for Product and its price
const catalog = new Map([
    ["Product A" , 20],
    ["Product B" , 40],
    ["Product C" , 50]
]);
    
console.log("\nProducts and their Price")

for(const [product, price] of catalog){
    console.log(`${product} : $${(price.toFixed(2))}`);
}

while(true){
    var quantityOfProductA = parseInt(prompt("Enter the quantity of product 'Product A': "));
    var quantityOfProductB = parseInt(prompt("Enter the quantity of product 'Product B': "));
    var quantityOfProductC = parseInt(prompt("Enter the quantity of product 'Product C': "));
    var isWrapped = prompt("Want to wrap as gift (yes / no)? ");

    // Validating the inputs so that negative values are not valid
    if ((quantityOfProductA >= 0 && quantityOfProductB >= 0 && quantityOfProductC >= 0) && (isWrapped.toLowerCase() == 'yes' || isWrapped.toLowerCase() == 'no')){
        break
    }
    else{
        console.log("Invalid inputs. Try again!")
    }  
}

const totalAmountOfProductA = catalog.get("Product A") * quantityOfProductA;
const totalAmountOfProductB = catalog.get("Product B") * quantityOfProductB;
const totalAmountOfProductC = catalog.get("Product C") * quantityOfProductC;

//  total amount in cart
const cartTotal = totalAmountOfProductA + totalAmountOfProductB + totalAmountOfProductC
//  total quantity of products
const totalQuantity = quantityOfProductA + quantityOfProductB + quantityOfProductC

const discountAmounts = applyDiscount();
// display the invoice
console.log("\nInvoice:")
display(discountAmounts)

File: shopping_cart.py
# discount 1 - flat_10_discount
def flat_10_discount():
    return 10.0

# discount 2 - bulk_5_discount
def bulk_5_discount(priceOfProduct):
    return priceOfProduct * (5 / 100)

# discount 3 - bulk_10_discount
def bulk_10_discount(cartTotal):
    return cartTotal * (10 / 100)

# discount 4 - tiered_50_discount
def tiered_50_discount(priceOfDiscountProduct):
    return priceOfDiscountProduct * (50/100)

def applyDiscount():
    # dictionary to store the discount values with its discount name. {discount name : discount price}
    discountAmounts = {}

    # check flat_10_discount
    if cartTotal > 200:
        flat_10_discount_amount = flat_10_discount()
        discountAmounts["flat_10_discount"] = flat_10_discount_amount
    # check bulk_5_discount
    if(quantityOfProductA > 10) or (quantityOfProductB > 10) or (quantityOfProductC > 10):
        bulk_5_discount_amount_A = 0
        bulk_5_discount_amount_B = 0
        bulk_5_discount_amount_C = 0
        if(quantityOfProductA > 10):
            bulk_5_discount_amount_A = bulk_5_discount(totalAmountOfProductA)
        if(quantityOfProductB > 10):
            bulk_5_discount_amount_B = bulk_5_discount(totalAmountOfProductB)
        if(quantityOfProductC > 10):
            bulk_5_discount_amount_C = bulk_5_discount(totalAmountOfProductC)
        bulk_5_discount_amount = bulk_5_discount_amount_A + bulk_5_discount_amount_B + bulk_5_discount_amount_C
        discountAmounts["bulk_5_discount"] = bulk_5_discount_amount
    # check bulk_10_discount
    if(totalQuantity > 20):
        bulk_10_discount_amount = bulk_10_discount(cartTotal)
        discountAmounts["bulk_10_discount"] = bulk_10_discount_amount
    # check tiered_50_discount
    if(totalQuantity > 30):
        tiered_50_discount_amount_A = 0
        tiered_50_discount_amount_B = 0 
        tiered_50_discount_amount_C = 0
        if(quantityOfProductA > 15):
            remainingQuantity = quantityOfProductA - 15
            tiered_50_discount_amount_A = (remainingQuantity * catalog.get("Product A")) - tiered_50_discount(remainingQuantity * catalog.get("Product A"))
        if(quantityOfProductB > 15):
            remainingQuantity = quantityOfProductB - 15
            tiered_50_discount_amount_B = (remainingQuantity * catalog.get("Product B")) - tiered_50_discount(remainingQuantity * catalog.get("Product B"))
        if(quantityOfProductC > 15):
            remainingQuantity = quantityOfProductC - 15
            tiered_50_discount_amount_C = (remainingQuantity * catalog.get("Product C")) - tiered_50_discount(remainingQuantity * catalog.get("Product C"))
        tiered_50_discount_amount = tiered_50_discount_amount_A + tiered_50_discount_amount_B + tiered_50_discount_amount_C
        discountAmounts["tiered_50_discount"] = tiered_50_discount_amount
    return discountAmounts

def giftWrap():
    if(isWrapped == "yes"):
        wrapAmount = totalQuantity * 1.0
    elif(isWrapped == "no"):
        wrapAmount = 0.0   
    return wrapAmount

def shipping():
    # 10units = 1 package
    unitsPerPackage = 10
    package = totalQuantity // unitsPerPackage
    remainingUnits = totalQuantity % unitsPerPackage
    if remainingUnits > 0 and remainingUnits <= unitsPerPackage:
        package = package + 1
    shippingFee = package * 5.0
    return shippingFee

def display(discountAmounts):
    print("\nProduct Name\tQuantity\tTotal Amount of that Product")
    print(f"Product A\t{quantityOfProductA}\t\t${totalAmountOfProductA:.2f}")
    print(f"Product B\t{quantityOfProductB}\t\t${totalAmountOfProductB:.2f}")
    print(f"Product C\t{quantityOfProductC}\t\t${totalAmountOfProductC:.2f}")
    print(f"\nSubTotal              : ${cartTotal:.2f}")
    # Minimum discount amount is beneficial to the customer
    beneficialDiscount = max(discountAmounts, key = discountAmounts.get)
    print(f"Discount Name Applied : {beneficialDiscount}")
    print(f"Discount Amount       : ${discountAmounts[beneficialDiscount]:.2f}")
    shipFee = shipping()
    giftWrapFee = giftWrap()
    print(f"Shipping Fee          : ${shipFee:.2f}")
    print(f"Gift Wrap Fee         : ${giftWrapFee:.2f}")
    totalAmount = cartTotal - discountAmounts[beneficialDiscount] + shipFee + giftWrapFee
    print(f"Total                 : ${totalAmount:.2f}")

# Defining a dictionary of catalog "Product Name : Price"
catalog = {
    "Product A" : 20.0,
    "Product B" : 40.0,
    "Product C" : 50.0
}

print("\nProducts and their Price")

for product, price in catalog.items():
    print(f"{product} : ${price:.2f}")

while(True):
    quantityOfProductA = int(input("Enter the quantity of product 'Product A': "))
    quantityOfProductB = int(input("Enter the quantity of product 'Product B': "))
    quantityOfProductC = int(input("Enter the quantity of product 'Product C': "))
    isWrapped = input("Want to wrap as gift (yes / no)? ")
    # Validating the inputs so that negative values are not valid
    if (quantityOfProductA >= 0 and quantityOfProductB >= 0 and quantityOfProductC >= 0) and (isWrapped.lower() == 'yes' or isWrapped.lower() == 'no'):
        break
    else:
        print("Invalid inputs. Try again!")

totalAmountOfProductA = catalog.get("Product A") * quantityOfProductA
totalAmountOfProductB = catalog.get("Product B") * quantityOfProductB
totalAmountOfProductC = catalog.get("Product C") * quantityOfProductC

# total amount in cart
cartTotal = totalAmountOfProductA + totalAmountOfProductB + totalAmountOfProductC

# total quantity of products
totalQuantity = quantityOfProductA + quantityOfProductB + quantityOfProductC
discountAmounts = applyDiscount()

# display the invoice
print("\nInvoice:")
display(discountAmounts)
